bson-transcoder/src/serializer.rs

```diff
@@ -0,0 +1,1056 @@
+                return hex;
+            }
+            // Binary: {"$binary": {"base64": "...", "subType": "..."}}
+            if let Some(serde_json::Value::Object(bin)) = obj.remove("$binary") {
+                if let Some(serde_json::Value::String(b64)) = bin.get("base64") {
+                    return b64.clone();
+                }
+            }
+            // Fallback: JSON-serialize the object
+            serde_json::to_string(&serde_json::Value::Object(obj)).unwrap_or_default()
+        }
+        other => serde_json::to_string(&other).unwrap_or_default(),
+    }
+}
+
+/// Stringify an _id value from JSON bytes for the root document.
+/// Parses the JSON and extracts string representation from common _id types.
+pub fn stringify_id(buf: &[u8]) -> String {
+    serde_json::from_slice(buf)
+        .map(id_to_string)
+        .unwrap_or_default()
+}
+
+pub struct SanitizingSerializeStruct<'a, W: Write> {
+    serializer: &'a mut SanitizingSerializer<W>,
+    first: bool,
+}
+
+impl<'a, W: Write> SerializeStruct for SanitizingSerializeStruct<'a, W> {
+    type Ok = ();
+    type Error = Error;
+
+    fn serialize_field<T: ?Sized + Serialize>(
+        &mut self,
+        key: &'static str,
+        value: &T,
+    ) -> Result<(), Self::Error> {
+        if !self.first {
+            self.serializer.writer.write_all(b",")?;
+        }
+        self.first = false;
+        serde_json::Serializer::new(&mut self.serializer.writer).serialize_str(key)?;
+        self.serializer.writer.write_all(b":")?;
+        value.serialize(&mut *self.serializer)
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        self.serializer.writer.write_all(b"}")?;
+        Ok(())
+    }
+}
+
+pub struct SanitizingSerializeStructVariant<'a, W: Write> {
+    serializer: &'a mut SanitizingSerializer<W>,
+    first: bool,
+}
+
+impl<'a, W: Write> SerializeStructVariant for SanitizingSerializeStructVariant<'a, W> {
+    type Ok = ();
+    type Error = Error;
+
+    fn serialize_field<T: ?Sized + Serialize>(
+        &mut self,
+        key: &'static str,
+        value: &T,
+    ) -> Result<(), Self::Error> {
+        if !self.first {
+            self.serializer.writer.write_all(b",")?;
+        }
+        self.first = false;
+        serde_json::Serializer::new(&mut self.serializer.writer).serialize_str(key)?;
+        self.serializer.writer.write_all(b":")?;
+        value.serialize(&mut *self.serializer)
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        self.serializer.writer.write_all(b"}}")?;
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn serialize_to_string<T: Serialize>(value: &T, is_root: bool) -> String {
+        let mut buf = Vec::new();
+        let mut serializer = SanitizingSerializer::new(&mut buf, is_root);
+        value.serialize(&mut serializer).unwrap();
+        String::from_utf8(buf).unwrap()
+    }
+
+    /// Test integration with actual BSON transcoding
```

Comment:
This comment doesn't really match the function, could probably just remove it since the function name is fine.